readthedocs/oauth/services/bitbucket.py

```diff
@@ -204,15 +215,64 @@ def setup_webhook(self, project):
                 headers={'content-type': 'application/json'}
             )
             if resp.status_code == 201:
+                recv_data = resp.json()
+                integration.provider_data = recv_data
+                integration.save()
                 log.info('Bitbucket webhook creation successful for project: %s',
                          project)
                 return (True, resp)
-        except RequestException:
+        # Catch exceptions with request or deserializing JSON
+        except (RequestException, ValueError):
             log.error('Bitbucket webhook creation failed for project: %s',
                       project, exc_info=True)
         else:
             log.error('Bitbucket webhook creation failed for project: %s',
                       project)
-            log.debug('Bitbucket webhook creation failure response: %s',
-                      resp.content)
+            try:
+                log.debug('Bitbucket webhook creation failure response: %s',
+                          resp.json())
+            except ValueError:
+                pass
+            return (False, resp)
+
+    def update_webhook(self, project, integration):
+        """Update webhook integration
+
+        :param project: project to set up webhook for
+        :type project: Project
+        :param integration: Webhook integration to update
+        :type integration: Integration
+        :returns: boolean based on webhook set up success, and requests Response object
+        :rtype: (Bool, Response)
+        """
+        session = self.get_session()
+        data = self.get_webhook_data(project, integration)
+        resp = None
+        try:
+            # Expect to throw KeyError here if provider_data is invalid
+            url = integration.provider_data['links']['self']['href']
+            resp = session.put(
+                url,
+                data=data,
+                headers={'content-type': 'application/json'}
+            )
+            if resp.status_code == 200:
+                recv_data = resp.json()
+                integration.provider_data = recv_data
+                integration.save()
+                log.info('Bitbucket webhook update successful for project: %s',
+                         project)
+                return (True, resp)
+        # Catch exceptions with request or deserializing JSON
+        except (KeyError, RequestException, ValueError):
+            log.error('Bitbucket webhook update failed for project: %s',
+                      project, exc_info=True)
+        else:
+            log.error('Bitbucket webhook update failed for project: %s',
+                      project)
+            try:
+                log.debug('Bitbucket webhook update failure response: %s',
+                          resp.json())
```

Comment:
Pretty sure I had a patch that removed the json() call because it blows up on non-JSON data. We should comment here if we don't want to capture non-JSON return data.